File: lib/src/agent_state.dart
/// Per-overlay-agent lifecycle state from AO (`type: agent_state`).
library;

/// Sticky per-agent readiness on a Reach session.
///
/// [pulling] is first-class (not a reason under [starting]) and is used only
/// for local Ollama ensure/pull.
enum AgentLifecycleState {
  down,
  starting,
  pulling,
  ready,
  busy,
  stopping,
}

AgentLifecycleState? parseAgentLifecycleState(String? raw) {
  switch ((raw ?? '').trim().toLowerCase()) {
    case 'down':
      return AgentLifecycleState.down;
    case 'starting':
      return AgentLifecycleState.starting;
    case 'pulling':
      return AgentLifecycleState.pulling;
    case 'ready':
      return AgentLifecycleState.ready;
    case 'busy':
      return AgentLifecycleState.busy;
    case 'stopping':
      return AgentLifecycleState.stopping;
    default:
      return null;
  }
}

String agentLifecycleStateWire(AgentLifecycleState state) {
  switch (state) {
    case AgentLifecycleState.down:
      return 'down';
    case AgentLifecycleState.starting:
      return 'starting';
    case AgentLifecycleState.pulling:
      return 'pulling';
    case AgentLifecycleState.ready:
      return 'ready';
    case AgentLifecycleState.busy:
      return 'busy';
    case AgentLifecycleState.stopping:
      return 'stopping';
  }
}

/// One `agent_state` frame from the engine.
class AgentStateUpdate {
  AgentStateUpdate({
    required this.agentProviderId,
    required this.state,
    this.reason,
    this.detail,
    this.model,
    this.progress,
    this.questionId,
    this.raw = const {},
  });

  factory AgentStateUpdate.fromJson(Map<String, dynamic> json) {
    final pid = (json['agentProviderId'] ?? json['agent_provider_id'] ?? '')
        .toString()
        .trim();
    final state = parseAgentLifecycleState(json['state']?.toString());
    if (state == null) {
      throw FormatException('unknown agent state: ${json['state']}');
    }
    final progress = json['progress'];
    return AgentStateUpdate(
      agentProviderId: pid,
      state: state,
      reason: json['reason']?.toString(),
      detail: json['detail']?.toString(),
      model: json['model']?.toString(),
      progress: progress is num ? progress.toDouble() : null,
      questionId: (json['questionId'] ?? json['question_id'])?.toString(),
      raw: Map<String, dynamic>.from(json),
    );
  }

  final String agentProviderId;
  final AgentLifecycleState state;
  final String? reason;
  final String? detail;
  final String? model;
  final double? progress;
  final String? questionId;
  final Map<String, dynamic> raw;

  @override
  String toString() =>
      'AgentStateUpdate($agentProviderId → ${agentLifecycleStateWire(state)})';
}

File: python/ao_reach/agent_state.py
"""Per-overlay-agent lifecycle state from AO (`type: agent_state`)."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any


class AgentLifecycleState(str, Enum):
    """Sticky per-agent readiness on a Reach session.

    ``PULLING`` is first-class (not a reason under ``STARTING``) and is used only
    for local Ollama ensure/pull.
    """

    DOWN = "down"
    STARTING = "starting"
    PULLING = "pulling"
    READY = "ready"
    BUSY = "busy"
    STOPPING = "stopping"


@dataclass(frozen=True)
class AgentStateUpdate:
    """One ``agent_state`` frame from the engine."""

    agent_provider_id: str
    state: AgentLifecycleState
    reason: str | None = None
    detail: str | None = None
    model: str | None = None
    progress: float | None = None
    question_id: str | None = None
    raw: dict[str, Any] | None = None

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> AgentStateUpdate:
        pid = str(
            data.get("agentProviderId")
            or data.get("agent_provider_id")
            or ""
        ).strip()
        raw_state = str(data.get("state") or "").strip().lower()
        try:
            state = AgentLifecycleState(raw_state)
        except ValueError as exc:
            raise ValueError(f"unknown agent state: {raw_state!r}") from exc
        progress = data.get("progress")
        qid = data.get("questionId")
        if qid is None:
            qid = data.get("question_id")
        return cls(
            agent_provider_id=pid,
            state=state,
            reason=(str(data["reason"]) if data.get("reason") is not None else None),
            detail=(str(data["detail"]) if data.get("detail") is not None else None),
            model=(str(data["model"]) if data.get("model") is not None else None),
            progress=float(progress) if isinstance(progress, (int, float)) else None,
            question_id=str(qid) if qid is not None else None,
            raw=dict(data),
        )
